README.md
## NAME
   M_uuid(3f) - a module of UUID (Universally Unique IDentifier) procedures
   (LICENSE:CUSTOM OPEN)
## SYNOPSIS
```text
       use m_uuid, only : generate_uuid
```
## QUOTE
####   __Remember you are unique, just like everyone else__.

## DESCRIPTION
   A universally unique identifier (UUID) is a 128-bit number used to
   identify information on computer systems. In particular, they are
   commonly used to create unique filenames and database and table keys.

   When generated according to the standard methods, UUIDs are for
   practical purposes unique, without depending for their uniqueness
   on a central registration authority or coordination between the
   parties generating them, unlike most other numbering schemes. While
   the probability that a UUID will be duplicated is not zero, it is
   close enough to zero to be negligible.

   Thus, anyone can create a UUID and use it to identify something with
   near certainty that the identifier does not duplicate one that has
   already been or will be created to identify something else. Information
   labeled with UUIDs by independent parties can therefore be later
   combined into a single database, or transmitted on the same channel,
   without needing to resolve conflicts between identifiers.

   Adoption of UUIDs and GUIDs is widespread. Many computing platforms
   provide support for generating them, and for parsing their textual
   representation.

   Note that several UUID types contain information that can be decyphered
   to recreate the creation time or the MAC address of the device that
   generated the UUID which has several uses for verifying when and
   where data was generated.

   [RFC 4122](https://tools.ietf.org/html/rfc4122)
   defines a Uniform Resource Name (URN) namespace for UUIDs.
   A UUID presented as a URN appears as follows:
```text
             urn:uuid:123e4567-e89b-12d3-a456-426655440000
```
       **-- Wikipedia**

## PROCEDURES
       * generate_uuid(version) ! generate 36-character UUID string


## BUILDING THE MODULE USING make(1)
     git clone https://github.com/urbanjost/M_uuid.git
     cd M_uuid/src
     # change Makefile if not using one of the listed compilers
     
     # for gfortran
     make clean
     make F90=gfortran gfortran
     
     # for ifort
     make clean
     make F90=ifort ifort

     # for nvfortran
     make clean
     make F90=nvfortran nvfortran

This will compile the Fortran module and basic example
program that exercise the routine.

## BUILD and TEST with FPM ![-](docs/images/fpm_logo.gif)

   Alternatively, download the github repository and build it with
   fpm ( as described at [Fortran Package Manager](https://github.com/fortran-lang/fpm) )

   ```bash
        git clone https://github.com/urbanjost/M_uuid.git
        cd M_uuid
        fpm run
        fpm run --example
        fpm test
   ```

   or just list it as a dependency in your fpm.toml project file.

```toml
        [dependencies]
        M_uuid        = { git = "https://github.com/urbanjost/M_uuid.git" }
```

## DOCUMENTATION

### USER

   - a simple index to the man-pages for the
   [routines](https://urbanjost.github.io/M_uuid/man3.html) 
   and [programs](https://urbanjost.github.io/M_uuid/man1.html) 
   in HTML form.

   - A single page that uses javascript to combine all the HTML
     descriptions of the manpages is at 
     [BOOK_M_uuid](https://urbanjost.github.io/M_uuid/BOOK_M_uuid.html).

   - [CHANGELOG](docs/CHANGELOG.md) provides a history of significant changes

### DEVELOPER
   - [ford(1) output](https://urbanjost.github.io/M_uuid/fpm-ford/index.html).
<!--
   - [doxygen(1) output](https://urbanjost.github.io/M_uuid/doxygen_out/html/index.html).
-->
   - [github action status](docs/STATUS.md) 
---
## PEDIGREE
 This is a modified version of generate_uuid(3f).  generate_uuid(3f)
 was originally derived from the xmlf90 codebase, (c) Alberto Garcia &
 Jon Wakelin, 2003-2004.  It also calls RNG routines from Scott Ladd
 <scott.ladd@coyotegulch.com>, and the libFoX modules. Although some
 sections have been replaced, generate_uuid(3f) was originally based on
 the libFoX version.

## REFERENCES ![-](docs/images/ref.gif)

   * [RFC-4122](https://tools.ietf.org/html/rfc4122)
   * [Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier)
   * [FOX](http://fortranwiki.org/fortran/show/FoX)
---


CHANGELOG.md
## M_uuid Changelog

The intent of this changelog is to keep everyone in the loop about
what's new in the M_uuid  project. It is a curated, chronologically ordered
list of notable changes including records of change such as bug fixes,
new features, changes, and relevant notifications.

---
**2021-02-10**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:
     initial release on github
---

<!--
   - [x] manpage
   - [x] demo program
   - [ ] unit test
### :orange_circle: DIFF:
       + renamed ADVICE(3f) to ALERT(3f)
### :green_circle: ADD:
       + advice(3f) was added to provide a standardized message format simply.
### :red_circle: FIX:
       + </bo> did not work on several terminal types, changed it to a more
         universally accepted value.
-->


test/test_suite_M_uuid.f90
module scr_uuid
use M_msg
use M_verify, only : unit_check_command, unit_check_keep_going, unit_check_level
use M_uuid, only : generate_uuid
private
public test_suite_m_uuid
contains
subroutine test_suite_m_uuid()
! this should contains tests for all public procedures in the module
   call test_generate_uuid()
end subroutine test_suite_m_uuid
!==================================================================================================================================!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!==================================================================================================================================!
subroutine test_generate_uuid()
use M_verify, only : unit_check, unit_check_start, unit_check_good, unit_check_bad, unit_check_done

!   This just checks that we can generate the various types of UUID
!   (without crashing) and checks that they have the correct syntax. We
!   could also check that the UUID changes for each call and I think there
!   is an additional check we could make within the UUID itself. But for
!   now this is enough.

character(len=36) :: uuid
   call unit_check_start('generate_uuid') ! start tests

   uuid = generate_uuid(0)
   call unit_check('generate_uuid',check_uuid(uuid).and.(uuid =='00000000-0000-0000-0000-000000000000'),msg='Version 0 '//uuid)

   uuid = generate_uuid(1)
   call unit_check('generate_uuid',check_uuid(uuid),msg='Version 1 '//uuid)

   uuid = generate_uuid(2)
   call unit_check('generate_uuid',uuid=='',msg='Version 2 (NOT IMPLEMENTED)')

   uuid = generate_uuid(3)
   call unit_check('generate_uuid',uuid=='',msg='Version 3 (NOT IMPLEMENTED)')

   uuid = generate_uuid(4)
   call unit_check('generate_uuid',check_uuid(uuid),msg='Version 4 '//uuid)

   uuid = generate_uuid(5)
   call unit_check('generate_uuid',uuid=='',msg='Version 5 (NOT IMPLEMENTED)')

   call unit_check('compare',exercise(1000000),msg='test for duplicates in 1000000 values')
   call unit_check_done('generate_uuid')
!==================================================================================================================================!
contains
!==================================================================================================================================!
function check_uuid(chars) result(lout)

! Return true if the string is permitted by the UUID BFN in RFC

character(len=*) :: chars
character(len=22), parameter :: hex = '0123456789abcdefABCDEF'
logical :: lout

   lout = (len_trim(chars) == 36)
   if (lout) then
       lout = lout.and.(verify(chars(1:8), hex) == 0)
       lout = lout.and.(verify(chars(9:9), '-') == 0)
       lout = lout.and.(verify(chars(10:13), hex) == 0)
       lout = lout.and.(verify(chars(14:14), '-') == 0)
       lout = lout.and.(verify(chars(15:18), hex) == 0)
       lout = lout.and.(verify(chars(19:19), '-') == 0)
       lout = lout.and.(verify(chars(20:23), hex) == 0)
       lout = lout.and.(verify(chars(24:24), '-') == 0)
       lout = lout.and.(verify(chars(25:36), hex) == 0)
   endif

end function check_uuid
!==================================================================================================================================!
end subroutine test_generate_uuid
!==================================================================================================================================!
function exercise(sz)
logical :: exercise
integer :: sz
character(len=36),allocatable :: uuid(:)
integer :: i,j
   exercise=.true.
   TYPES: do j=1,4,3
      if(allocated(uuid))deallocate(uuid)
      allocate(uuid(sz))
      do i=1,sz
         uuid(i)=generate_uuid(j)
      enddo
      call sort_shell_strings_lh(uuid)
      call unique_strings(uuid,icount)

      if(icount.ne.size(uuid))then
<<<<<<< HEAD
=======
         write(*,*)'error: duplicates found in ',size(uuid),' values'
>>>>>>> e988d53f7dafcb9b0a48e48f9d5355cfbfd3e4fa
         exercise=.false.
         exit TYPES
      endif
   enddo TYPES
end function exercise
!==================================================================================================================================!
subroutine sort_shell_strings_lh(lines)
! ident_5="@(#)M_sort::sort_shell_strings_lh(3fp):sort strings(a-z) over specified field using shell sort"
character(len=*) :: lines(:)
   character(len=:),allocatable :: ihold
   integer           :: n, igap, i,j,k, jg
   n=size(lines)
   if(n.gt.0)then
      allocate(character(len=len(lines(1))) :: ihold)
   else
      ihold=''
   endif
   igap=n
   INFINITE: do
      igap=igap/2
      if(igap.eq.0) exit INFINITE
      k=n-igap
      i=1
      INNER: do
         j=i
         INSIDE: do
            jg=j+igap
            if(lle(lines(j),lines(jg)))exit INSIDE
            ihold=lines(j)
            lines(j)=lines(jg)
            lines(jg)=ihold
            j=j-igap
            if(j.lt.1) exit INSIDE
         enddo INSIDE
         i=i+1
         if(i.gt.k) exit INNER
      enddo INNER
   enddo INFINITE
end subroutine sort_shell_strings_lh

subroutine unique_strings(array,ivals)
character(len=*),intent(inout),allocatable  :: array(:)
integer,intent(out)                         :: ivals
   integer                                  :: i,isize
   isize=size(array)
   if(isize.ge.2)then
      ivals=1
      do i=2,isize
        if(array(i).ne.array(i-1))then
           ivals=ivals+1
<<<<<<< HEAD
        else
           write(*,*)'<ERROR> at ',i,' and ', i+1,' duplicates ',array(i)
=======
           array(ivals)=array(i)
>>>>>>> e988d53f7dafcb9b0a48e48f9d5355cfbfd3e4fa
        endif
      enddo
   else
      ivals=isize
   endif
end subroutine unique_strings
!==================================================================================================================================!
end module scr_uuid
!==================================================================================================================================!
program runtest
use M_msg
use M_verify, only : unit_check_command, unit_check_keep_going, unit_check_level
use scr_uuid
implicit none
   unit_check_command=''
   unit_check_keep_going=.true.
   unit_check_level=0
   call test_suite_M_uuid()
end program runtest
!==================================================================================================================================!


app/unique.f90
program unique ! this is a utility program. It is typically built using ccall(1).
use M_CLI2,                       only : set_args, lget, sgets, iget, unnamed
use M_uuid,                       only : generate_uuid
use,intrinsic :: iso_fortran_env, only : stdin=>input_unit, stdout=>output_unit, stderr=>error_unit
implicit none
integer :: version
integer :: repeat
integer :: i,j
character(len=10),allocatable :: methods(:)
character(len=:),allocatable  :: prefix
character(len=:),allocatable  :: help_text(:)
character(len=:),allocatable  :: version_text(:)

   ! define arguments, default values and crack command line
   call setup()
   call set_args('--method:m "4" --urn:U F --repeat:r 1',help_text,version_text)
   methods=[character(len=10) :: sgets('method'),unnamed]        ! get value of command line argument -method or unnamed strings
   if(size(methods).eq.0)methods=["4"]
   repeat=iget('repeat')                                         ! get value of command line argument -repeat
   prefix=merge('urn:uuid:','         ',lget('urn'))             ! get value of command line argument -urn
   prefix=trim(prefix)

   do i=1,size(methods)
      select case(methods(i))
      case(   '0','nil');    version=0
      case(   '1','time');   version=1
      case('','4','random'); version=4
      case default;          version=4
         write(stderr,'(*(g0,1x))')'*unique* unknown method',methods
         stop 1
      endselect
      do j=1,repeat
         write(*,'(2a)')prefix,generate_uuid(version)
      enddo
   enddo

contains
subroutine setup()

! @(#)help_usage(3f): sets help information

version_text=[ CHARACTER(LEN=128) :: &
'@(#)PRODUCT:        GPF (General Purpose Fortran) utilities and examples>',&
'@(#)PROGRAM:        unique(1)>',&
'@(#)DESCRIPTION:    output a UUID (Universally Unique ID)>',&
'@(#)VERSION:        1.0, 20180427>',&
'@(#)AUTHOR:         John S. Urban>',&
'@(#)HOME PAGE:      http://www.urbanjost.altervista.org/index.html>',&
'@(#)COMPILED:       Tue, Feb 9th, 2021 9:19:16 PM>',&
'']

help_text=[ CHARACTER(LEN=128) :: &
'NAME',&
'    unique(1f) - [FUNIX] generate a UUID (Universally Unique ID) string',&
'                  per RFC 4122',&
'    (LICENSE:PD)',&
'',&
'SYNOPSIS',&
'    unique [[ --method] NAME][ -urn][ -repeat N]]|[ --help|--version]',&
'',&
'DESCRIPTION',&
'   unique(3f) generates UUID strings according to the RFC 4122',&
'   standard.',&
'',&
'   A universally unique identifier (UUID) is a 128-bit number used to',&
'   identify information in computer systems. When generated according',&
'   to standard methods UUIDs are for practical purposes unique.',&
'',&
'   Standard methods 0,1, and 4 are supported as described in RFC 4122.',&
'',&
'   UUID strings are particularly useful as keys for relational database',&
'   entries, and for building unique temporary file names (especially in',&
'   cross-mounted filesystems that more than one OS is utilizing).',&
'',&
'OPTIONS',&
'    --method NAME  Select the UUID version type. Supported methods are',&
'                   nil|0, random|4, time|1.',&
'',&
'                   0. Nil UUID (ie. ''''00000000-0000-0000-0000-000000000000'''')',&
'                   1. time-based UUID',&
'                   2. Not implemented',&
'                   3. Not implemented',&
'                   4. pseudo-RNG(Random Number Generator) based',&
'                   5. Not implemented',&
'',&
'    --urn       RFC 4122 defines a Uniform Resource Name (URN)',&
'                namespace for UUIDs. IE., the output is',&
'                prefixed with "urn:uuid:".',&
'',&
'    --repeat N  Number of UUID strings to generate',&
'',&
'    --help      display this help and exit',&
'',&
'    --version   output version information and exit',&
'',&
'EXAMPLES',&
'  Sample commands',&
'',&
'   unique',&
'   4bb8051e-4af3-11e8-6603-4254ffee9a14',&
'',&
'   unique -urn',&
'   urn:uuid:e9fd7cab-69f2-4cd6-4b5e-d54b9fbf617a',&
'',&
'   unique -method time -repeat 4',&
'   f2a2faf0-833a-11e9-7373-5eb4cfd7e237',&
'   f2a2faf0-833a-11e9-7373-afbb9f7b9100',&
'   f2a2faf0-833a-11e9-7373-cde3ffff3681',&
'   f2a2faf0-833a-11e9-7373-271cfbfd42bc',&
'',&
'AUTHOR',&
'   John S. Urban',&
'LICENSE',&
'   Public Domain',&
'']
! NAME
!     unique(1f) - [FUNIX] generate a UUID (Universally Unique ID) string
!                   per RFC 4122
!     (LICENSE:PD)
! 
! SYNOPSIS
!     unique [[ --method] NAME][ -urn][ -repeat N]]|[ --help|--version]
! 
! DESCRIPTION
!    unique(3f) generates UUID strings according to the RFC 4122
!    standard.
! 
!    A universally unique identifier (UUID) is a 128-bit number used to
!    identify information in computer systems. When generated according
!    to standard methods UUIDs are for practical purposes unique.
! 
!    Standard methods 0,1, and 4 are supported as described in RFC 4122.
! 
!    UUID strings are particularly useful as keys for relational database
!    entries, and for building unique temporary file names (especially in
!    cross-mounted filesystems that more than one OS is utilizing).
! 
! OPTIONS
!     --method NAME  Select the UUID version type. Supported methods are
!                    nil|0, random|4, time|1.
! 
!                    0. Nil UUID (ie. ''00000000-0000-0000-0000-000000000000'')
!                    1. time-based UUID
!                    2. Not implemented
!                    3. Not implemented
!                    4. pseudo-RNG(Random Number Generator) based
!                    5. Not implemented
! 
!     --urn       RFC 4122 defines a Uniform Resource Name (URN)
!                 namespace for UUIDs. IE., the output is
!                 prefixed with "urn:uuid:".
! 
!     --repeat N  Number of UUID strings to generate
! 
!     --help      display this help and exit
! 
!     --version   output version information and exit
! 
! EXAMPLES
!   Sample commands
! 
!    unique
!    4bb8051e-4af3-11e8-6603-4254ffee9a14
! 
!    unique -urn
!    urn:uuid:e9fd7cab-69f2-4cd6-4b5e-d54b9fbf617a
! 
!    unique -method time -repeat 4
!    f2a2faf0-833a-11e9-7373-5eb4cfd7e237
!    f2a2faf0-833a-11e9-7373-afbb9f7b9100
!    f2a2faf0-833a-11e9-7373-cde3ffff3681
!    f2a2faf0-833a-11e9-7373-271cfbfd42bc
! 
! AUTHOR
!    John S. Urban
! LICENSE
!    Public Domain
end subroutine setup
end program unique
